### PR TITLE
Case insensitive filter and copy code.

### PR DESCRIPTION
**Pull request by fakoua**
- Make filter in RegistryIndex case insensitive
- Allow Copy in code block

### CODE REVIEW

**Comment by lucacasonato** [comment]
:+1:
Could you maybe put the code copy at the end of the text field and remove the extra space at the start of the codeblock? I think the copy button should appear when you hover over the textfield, floating in the top right.

File: src/component/CodeBlock.tsx [hunk 1]
```diff
@@ -13,18 +17,79 @@ darkTheme["hljs-selection"] = {
...
-    <SyntaxHighlighter
-      style={darkMode ? darkTheme : lightTheme}
-      language={props.language || "js"}
-      showLineNumbers={props.showLineNumbers || false}
-      wrapLines={true}
-      lineProps={props.lineProps}
-    >
-      {props.value}
-    </SyntaxHighlighter>
+    <Container
```
Comment (lucacasonato) [inline src/component/CodeBlock.tsx]: This container adds an extra margin on the left and right - I don't think we want that.

File: src/component/CodeBlock.tsx [hunk 2]
```diff
@@ -13,18 +17,79 @@ darkTheme["hljs-selection"] = {
...
+      newState = Object.assign({}, state)
+      newState.copied = false;
+      setState(newState)
+    }, 300)
+  }
+
+  const [state, setState] = React.useState({
+    showCopy: false,
+    copied: false,
+  });
```
Comment (lucacasonato) [inline src/component/CodeBlock.tsx]: Could you maybe replace this with two state objects to make the state handling slightly less verbose - you can then get rid of all of the `Object.assign`s.

```typescript
const [showCopy, setShowCopy] = React.useState(false);
const [copied, setCopied] = React.useState(false)
```

**Comment by fakoua** [comment]
Feedback fixed!
Thanks

**Comment by lucacasonato** [comment]
Looks great! Is it possible to add a timeout for fade-out? So the copy button will stick around for half a second before fading out? Also I think a slightly shorter fade in would work better.

**Comment by fakoua** [comment]
Fixed, thanks

**Comment by lucacasonato** [comment]
Great! Just one more nitpick - could you remove the fade in and just keep the fade out? Also hovering over the code block, then leaving and then entering again makes the copy button disappear.

This should fix both things:

```typescript
  function onContainerEnter() {
    clearTimeout(timerState.timerId);
    setShowCopy({showCopy: true});
  }

  function onContainerLeave() {
    clearTimeout(timerState.timerId);
    let timerId = setTimeout(() => {
      setShowCopy({showCopy: false});
    }, 500);
    setTimer({timerId: timerId})
  }
```

**Comment by fakoua** [comment]
Many thanks, yes entering and leaving was buggy.

Issue fixed.

**Comment by ry** [comment]
This seems like a lot of code for what it’s achieving...

**Comment by fakoua** [comment]
@ry
You are right, could be simplified using css in some cases.

**Comment by fakoua** [comment]
@ry ,
It shows all the fork sync since started, the affected files one 2:
1- RegistryIndex.js:
```ts
name.toLowerCase().includes(query.toLowerCase()) ||
      (DATABASE[name].desc || "").toLowerCase().includes(query.toLowerCase())
```
2- CodeBlock.tsx:
```ts
const useStyles = makeStyles(theme => ({
  copyButton: {
    position: 'relative',
    right: '5px',
    float: 'right'
  },
  hidden: {
    display: 'none'
  },
  container: {
    position: 'relative',
    paddingLeft: '0',
    paddingRight: '0'
  },
  checkIcon: {
    color: 'rgb(80, 161, 79)'
  }
}));

...

const [timerState, setTimer] = React.useState({
    timerId: 0,
  });

  const [showCopyState, setShowCopy] = React.useState({
    showCopy: false,
  });
  
  const [copiedState, setCopied] = React.useState({
    copied: false,
  });

  function onContainerEnter() {
    clearTimeout(timerState.timerId);
    setShowCopy({showCopy: true});
  }

  function onContainerLeave() {
    clearTimeout(timerState.timerId);
    let timerId = setTimeout(() => {
      setShowCopy({showCopy: false});
    }, 500);
    setTimer({timerId: timerId})
  }

  function onCopy() {
    setCopied({copied: true});
    setTimeout(() => {
      setCopied({copied: false});
    }, 1300)
  }

<Container
      onMouseEnter={onContainerEnter}
      onMouseLeave={onContainerLeave}
      className={classes.container}>
      <Zoom in={showCopyState.showCopy || copiedState.copied} style={{ transitionDelay: '50ms' }}>
        <CopyToClipboard
          text={props.value}
          onCopy={onCopy}>
          <IconButton .........

```

**Comment by lucacasonato** [comment]
> This seems like a lot of code for what it’s achieving...

@ry I do think it is a really important feature though. Being able to copy like this is very, very handy.

---

@fakoua You can reduce the line count some more by replacing the state management to use direct primitive types and not objects.

```typescript
  const [timer, setTimer] = React.useState(0);
  const [showCopy, setShowCopy] = React.useState(false);
  const [copied, setCopied] = React.useState(false);

  function onContainerEnter() {
    clearTimeout(timer);
    setShowCopy(true);
  }

  function onContainerLeave() {
    clearTimeout(timer);
    let timerID = setTimeout(() => {
      setShowCopy(false);
    }, 500);
    setTimer(timerID)
  }

  function onCopy() {
    setCopied(true);
    setTimeout(() => {
      setCopied(false);
    }, 1300)
  }
```

**Comment by fakoua** [comment]
> > This seems like a lot of code for what it’s achieving...
> 
> @ry I do think it is a really important feature though. Being able to copy like this is very, very handy.
> 
> @fakoua You can reduce the line count some more by replacing the state management to use direct primitive types and not objects.
> 
> 
>   const [timer, setTimer] = React.useState(0);
>   const [showCopy, setShowCopy] = React.useState(false);
>   const [copied, setCopied] = React.useState(false);
> 
>   function onContainerEnter() {
>     clearTimeout(timer);
>     setShowCopy(true);
>   }
> 
>   function onContainerLeave() {
>     clearTimeout(timer);
>     let timerID = setTimeout(() => {
>       setShowCopy(false);
>     }, 500);
>     setTimer(timerID)
>   }
> 
>   function onCopy() {
>     setCopied(true);
>     setTimeout(() => {
>       setCopied(false);
>     }, 1300)
>   }

- Code cleanup, 
Thanks

**Comment by ry** [comment]
I’m away from my computer the next two days - I’ll review in detail when I’m back.

**Comment by ry** [comment]
This seems to have two unrelated bits. I wonder if they could be split into two PRs?

**Comment by lucacasonato** [comment]
@fakoua Maybe remove the case insensitive filtering from this PR and just keep the copy button in this PR. Then you can open an other PR for the case insensitive filtering.

**Comment by fakoua** [comment]
Hi @ry , @lucacasonato 

This PR contains only the copy code feature, once merged I will PR the case insensitive feature, 
Many thanks.

**Pull request merged**